Adding new TSG for troubleshooting AKS node auto-repair errors (#1629)
MicrosoftDocs/SupportArticles-docs: support/azure/azure-kubernetes/availability-performance/node-auto-repair-errors.md

---
title: Troubleshoot common node auto-repair errors
description: Troubleshoot scenarios where node auto-repair returns an error code when trying to repair your NotReady node.
ms.date: 10/01/2024
ms.reviewer: 
ms.service: azure-kubernetes-service
#Customer intent: As an Azure Kubernetes user, I want to make sure the automatic repair actions from AKS node auto-repair do not cause any impacts on my applications or cluster health.
ms.custom: sap:Node/node pool availability and performance
---
# Troubleshoot common node auto-repair errors

> When AKS detects a node which reports the NotReady status for more than 5 minutes, we will attempt to automatically repair your node. Node auto-repair is a best-effort service and does not guarantee that your node will be restored to a healthy state. Learn more about the [node auto-repair process](/azure/aks/node-auto-repair).

Author: How do you suggest to word this without saying "AKS" twice in the same sentence? Right now it will be: "When AKS detects a node which reports the NotReady status for more than 5 minutes, AKS will attempt to automatically repair your node."

## Prerequisites
To determine what type of node auto-repair error has occured, you will need to look for the following Kubernetes event: 
"Node auto-repair reboot/reimage/redeploy action failed due to an operation failure: [error code]."

> [!NOTE]
> In most cases, node auto-repair errors will not cause any impact to your cluster or applications. When encountering a node auto-repair error, we recommend that you attempt to repair the node yourself by following the instructions here: [Basic troubleshooting of Node Not Ready failures](./node-not-ready-basic-troubleshooting.md). If you are unable to restore your node back to a Succeeded status and are noticing persistent errors reported by node auto-repair, you may be encountering a rare issue and should reach out to Azure support.

## Common error codes
The table below contains the most common node auto-repair errors.

| Error code | Causes & Solution |
|---|---|

> | ARM ErrorCode: VMExtensionProvisioningError | One or more VM extensions failed to be provisioned on the VM. Read more on possible error types and troubleshooting steps at [Troubleshoot the ERR_VHD_FILE_NOT_FOUND error code (124)](../create-upgrade-delete/error-code-vhdfilenotfound.md). This may also occur as a result of custom script extension (CSE) errors on the node. Learn more at [Troubleshoot node not ready failures caused by CSE errors](./node-not-ready-custom-script-extension-errors.md). |

Reviewer: this TSG section might be better? https://learn.microsoft.com/en-us/troubleshoot/azure/azure-kubernetes/create-upgrade-delete/troubleshoot-aks-cluster-creation-issues#view-resources-in-the-azure-portal

That will lead to the actual VM extension provisioning error. The TSG linked to here seems to assume this is a node provisioning CSE error, and one of several specific error codes, when this could be due to a different extension.

| ARM ErrorCode: InvalidParameter | This occurs when the node auto-repair process requests a VM that no longer exists. [more guidance] |
| scaleSetNameAndInstanceIDFromProviderID failed | This is an internal issue where the node may not have been provisioned correctly. If you notice any new symptoms or interruptions on your node, please reach out to Azure support. |
| ManagedIdentityCredential authentication failed | This is likely due to an internal issue where your nodes were not initialized properly. If you notice any new symptoms or interruptions on your node, please reach out to Azure support. |

> | ARM ErrorCode: VMRedeploymentFailed | There was an error with redeploying your node. This may cause your nodepool to enter a Failed state. Read more on potential causes and next steps at [Troubleshoot Azure Kubernetes Service clusters or nodes in a failed state](./cluster-node-virtual-machine-failed-state.md#scenario-3-node-pool-is-in-a-failed-state). |

Reviewer 2: ```suggestion
| ARM ErrorCode: VMRedeploymentFailed | There was an error with redeploying your node. This may cause your nodepool to enter a failed state. Read more on potential causes and next steps at [Troubleshoot Azure Kubernetes Service clusters or nodes in a failed state](./cluster-node-virtual-machine-failed-state.md#scenario-3-node-pool-is-in-a-failed-state). |
```

> | ARM ErrorCode: TooManyVMRedeploymentRequests | This error occurs when your cluster exceeds the internal limit for VM redeployment requests. Redeploy is one of the node auto-repair actions, which means that we are unable to repair your node using redeploy actions. We recommend that you  troubleshoot your node's NotReady status to understand the root cause and resolve the node health issue. Read more at [Basic troubleshooting of Node Not Ready failures](./node-not-ready-basic-troubleshooting.md). |

Reviewer 2: ```suggestion
| ARM ErrorCode: TooManyVMRedeploymentRequests | This error occurs when your cluster exceeds the internal limit for VM redeployment requests. Redeploy is one of the node auto-repair actions, so this error indicates that we are unable to repair your node using redeploy actions. We recommend that you  troubleshoot your node's NotReady status to understand the root cause and resolve the node health issue. Read more at [Basic troubleshooting of Node Not Ready failures](./node-not-ready-basic-troubleshooting.md). |
```

| ARM ErrorCode: OutboundConnectivityNotEnabledOnVMSS | This error occurs when your node or overall Virtual Machine Scale Set (VMSS) does not have outbound access enabled. Enable secure outbound access for your VMSS by using a method that's best suited for your application. Read more about the methods at [OutboundConnectivityNotEnabledOnVM. No outbound connectivity configured for virtual machine](../../virtual-machine-scale-sets/deploy/vmss-outbound-connectivity-not-enabled.md#solution). |

## Best practices for monitoring

> - By default, AKS stores Kubernetes events from the past 1 hour. We recommend for you to enable [Container Insights](/azure/azure-monitor/containers/kubernetes-monitoring-enable#enable-container-insights) to store events for up to 90 days. Enabling Container Insights will also allow you to [query events and configure alerts](/azure/aks/events#automating-event-notifications) to quickly detect node auto-repair errors.

Reviewer 2: ```suggestion
- By default, AKS stores Kubernetes events from the past hour. We recommend that you enable [Container Insights](/azure/azure-monitor/containers/kubernetes-monitoring-enable#enable-container-insights) so that you can store events for up to 90 days. Enabling Container Insights will also allow you to [query events and configure alerts](/azure/aks/events#automating-event-notifications) to quickly detect node auto-repair errors.
```

## Common error codes
The table below contains the most common node auto-repair errors.

| Error code | Causes & Solution |
|---|---|
| ARM ErrorCode: VMExtensionProvisioningError | One or more VM extensions failed to be provisioned on the VM. Read more on possible error types and troubleshooting steps at [Troubleshoot the ERR_VHD_FILE_NOT_FOUND error code (124)](https://learn.microsoft.com/en-us/troubleshoot/azure/azure-kubernetes/create-upgrade-delete/error-code-vhdfilenotfound). This may also occur as a result of custom script extension (CSE) errors on the node. Learn more at [Troubleshoot node not ready failures caused by CSE errors](https://learn.microsoft.com/en-us/troubleshoot/azure/azure-kubernetes/availability-performance/node-not-ready-custom-script-extension-errors). |

> | ARM ErrorCode: InvalidParameter | This occurs due to a parameter causes errors when new nodes are created for an AKS cluster. Read more on causes and solution at [Troubleshoot the InvalidParameter error code](https://learn.microsoft.com/en-us/troubleshoot/azure/azure-kubernetes/create-upgrade-delete/error-code-invalidparameter#cause). |

Reviewer: This might be confusing since customers did not make the request.

Generally, the remediator service requested a VM that no longer exists.

Author: Sounds good. Is there anything else we can say here about the customer impact - is this typically due to quota or capacity issues? And will remediator request a different VM or further redeployments will just fail?

Reviewer: I've mentioned elsewhere, it can be spot VMs which get preempted frequently.
Otherwise it might have been deleted by cluster-autoscaler or something.

ms.custom: sap:Node/node pool availability and performance
---
# Troubleshoot common node auto-repair errors

When AKS detects a node which reports the NotReady status for more than 5 minutes, we will attempt to automatically repair your node. Learn more about the node auto-repair process [here](https://learn.microsoft.com/en-us/azure/aks/node-auto-repair).

During this process, AKS will initiate reboot, reimage, and redeploy actions on your unhealthy node. These repair actions may be unsuccessful due to an underlying cause, resulting in an error code. This article will discuss common errors with their potential causes and next steps, as well as best practices for monitoring node auto-repair.

## Prerequisites
To determine what type of node auto-repair error has occured, you will need to look for the following Kubernetes event: 
"Node auto-repair [reboot/reimage/redeploy] action failed due to an operation failure: [error code]."

## Common error codes
The table below contains the most common node auto-repair errors.

| Error code | Potential causes | Next steps |
|---|---|---|

> | ClientSecretCredential authentication failed |  |  |

Reviewer: as mentioned on teams, I think we can remove this one and add some other top errors. Since this is an issue with misclassifying MSI clusters which there is a fix rolling out for.

> | ARM ErrorCode: VMExtensionProvisioningError |  |  |

Reviewer: probably relevant much of the time: https://learn.microsoft.com/en-us/troubleshoot/azure/azure-kubernetes/availability-performance/node-not-ready-custom-script-extension-errors

> | ARM ErrorCode: InvalidParameter |  |  |

Reviewer: I think this is mostly an issue with spot VM node objects still existing for a little while after the VMs are preempted.

> | scaleSetNameAndInstanceIDFromProviderID failed |  |  |

Reviewer: seems to be uninitialized nodes

> | ManagedIdentityCredential authentication failed |  |  |

Reviewer 3: Looks good, as long as we have some common errors , next steps. This is good to go.

## Common error codes
The table below contains the most common node auto-repair errors.

| Error code | Causes & Solution |
|---|---|
| VMExtensionProvisioningError | One or more VM extensions failed to be provisioned on the VM. Read more on possible error types and troubleshooting steps at [Troubleshoot the ERR_VHD_FILE_NOT_FOUND error code (124)](../create-upgrade-delete/error-code-vhdfilenotfound.md). To determine the exact VM extension provisioning error on your node, read more at [View error details in the Azure portal](../create-upgrade-delete/troubleshoot-aks-cluster-creation-issues.md#view-resources-in-the-azure-portal). |
| InvalidParameter | This is an internal error which occurs when the node auto-repair process requests a VM that no longer exists. If you notice any new symptoms or interruptions on your node, please reach out to Azure support. |

> | scaleSetNameAndInstanceIDFromProviderID failed | This is an internal issue where the node may not have been provisioned correctly. If you notice any new symptoms or interruptions on your node, please reach out to Azure support. |

Reviewer: Just remembered this and "ManagedIdentityCredential authentication failed" are not going to show up unless the customer is seeing a Diagnose and Solve detector with remediator service logs, since these are not ARM error codes. We don't have that right now.
